velox/functions/sparksql/specialforms/MakeDecimal.cpp

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/functions/sparksql/specialforms/MakeDecimal.h"
+#include "velox/expression/ConstantExpr.h"
+
+namespace facebook::velox::functions::sparksql {
+namespace {
+template <typename T, bool nullOnOverflow>
+class MakeDecimalFunction final : public exec::VectorFunction {
+ public:
+  explicit MakeDecimalFunction(uint8_t resultPrecision)
+      : resultPrecision_(resultPrecision) {}
+
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args, // Not using const ref so we can reuse args
+      const TypePtr& outputType,
+      exec::EvalCtx& context,
+      VectorPtr& resultRef) const final {
+    VELOX_USER_CHECK_EQ(args.size(), 2, "Make_decimal expects two arguments.");
+    VELOX_USER_CHECK(
+        args[0]->isConstantEncoding() || args[0]->isFlatEncoding(),
+        "Single-arg deterministic functions receive their only argument as flat or constant vector.");
+    context.ensureWritable(rows, outputType, resultRef);
+    resultRef->clearNulls(rows);
+    auto result = resultRef->asUnchecked<FlatVector<T>>()->mutableRawValues();
+    int128_t bound = DecimalUtil::kPowersOfTen[resultPrecision_];
+    if (args[0]->isConstantEncoding()) {
+      // Fast path for constant vectors.
+      auto constant =
+          args[0]->asUnchecked<ConstantVector<int64_t>>()->valueAt(0);
+      if constexpr (std::is_same_v<T, int64_t>) {
+        bool overflow = constant <= -bound || constant >= bound;
+        rows.applyToSelected([&](vector_size_t row) {
+          if (overflow) {
```

Comment:
Can we move the check outside of the loop?

```
if (overflow) {
   if (nullOnOverflow) {
      result->addNulls(rows);
   } else {
      context.setErrors(rows, ...);
   }
} else {
   rows.applyToSelected([&](vector_size_t row) {
      rawResults[row] = constant;
   }
}
```

Reply:
Yes, you're right. Fixed.